application/backend/src/business/services/australian-genomics/redcap-import-application-service.ts

```diff
@@ -100,9 +115,78 @@ export class RedcapImportApplicationService {
         }
       }
 
-      // loop through the members (users) mapping to users that we know about
-      // TODO: create new users entries for those we don't know yet
-      // const memberToUserMap: { [uri: string]: string } = {};
+      let pi: ApplicationUser;
+
+      // the PI details are literally a duplicate of the applicant OR
+      // we grab them from elsewhere in the CSV
+      if (newApplication.daf_applicant_pi_yn === "1") {
+        pi = {
+          email: newApplication.daf_applicant_email,
+          displayName: newApplication.daf_applicant_name,
+          institution: newApplication.daf_applicant_institution,
+          role: "PI",
+        };
+      } else {
+        pi = {
+          email: newApplication.daf_pi_email,
+          displayName: newApplication.daf_pi_name,
+          institution:
+            newApplication.daf_pi_institution_same === "1"
+              ? newApplication.daf_applicant_institution
+              : newApplication.daf_pi_institution,
+          role: "PI",
+        };
+      }
+
+      const otherResearchers: ApplicationUser[] = [];
+
+      if (newApplication.daf_collab_num) {
+        for (
+          let otherResearchCount = 0;
+          otherResearchCount < parseInt(newApplication.daf_collab_num);
+          otherResearchCount++
+        ) {
+          const siteNumber = (otherResearchCount + 1).toString();
+          const isHoldingData =
+            (newApplication as any)["daf_data_house_site" + siteNumber] === "1";
+          const r: ApplicationUser = {
+            email: (newApplication as any)[
+              "daf_contact_email_site" + siteNumber
+            ],
+            displayName: (newApplication as any)[
+              "daf_contact_site" + siteNumber
+            ],
+            institution: (newApplication as any)[
+              "daf_institution_site" + siteNumber
+            ],
```

Comment:
Doesn't the type checker complain that possibly `undefined` values are getting assigned to things which should always be `string`s?

Instead of using `as any`, it'd be safer to define a utility function which gets an object's property if it exists or raises an exception otherwise. Then we'll get something like this:

```suggestion
            institution: getOrRaise(
              newApplication,
              "daf_institution_site" + siteNumber),
```

I was half-expecting `lodash` to have a function like that, but I couldn't see one. (The library follows a functional style, so maybe they're ideologically opposed to making functions not be total by raising exceptions.)

Reply:
Good thinking - I think I possibly need to add a general check on the validity of the resulting ApplicationUser (because we also need to protect against CSV columns that are empty strings for instance). So I'll add that in - and that will pick up any 'undefined' as well)